File: Train/CLRS/merge_sort.py
import math


def main():
    array = [int(_) for _ in input().strip().split()]

    res = merge_sort(array)
    print(res)


def merge_sort(array):
    if len(array) == 1:
        return array
    else:
        mid = len(array) // 2
        if mid:
            left = merge_sort(array[:mid])
            right = merge_sort(array[mid:])
            return merge(left, right)


def merge(a, b):
    i = 0
    j = 0
    a = a + [math.inf]
    b = b + [math.inf]
    res = []
    while i < len(a)-1 or j < len(b)-1:
        if a[i] < b[j]:
            res += [a[i]]
            i += 1
        else:
            res += [b[j]]
            j += 1
    return res


if __name__ == '__main__':
    main()

File: Train/CPP/CLRS/merge_sort.cpp
//
// Created by Lídia Maria Carvalho de Freitas on 08/03/2018.
//
#include <vector>
#include <iostream>


std::vector<int>& merge_sort(std::vector<int>&);
std::vector<int>& merge(std::vector<int>& array1, std::vector<int>& array2);

int main(){
  std::vector<int> array;
  int x;

  while (std::cin >> x) array.push_back(x);

  array = merge_sort(array);


}

std::vector<int>& merge_sort(std::vector<int>& array,long i, long j){
  if (i < j){
    long mid = i + (j - i) / 2;
    std::vector<int> left  = merge_sort(array, i, mid);
    std::vector<int> right = merge_sort(array, mid+1,j);
    return merge(left, right);
  }
}
std::vector<int>& merge(std::vector<int> &array1, std::vector<int> &array2) {
  int i=0,j=0;
  std::vector<int> result;

  while (i < array1.size() && j < array2.size()){
    if (i == array1.size()){
      result.push_back(array2[j]);
      ++j;
    }
    else if (j == array2.size()){
      result.push_back(array1[i]);
      ++i;
    }
    else {
      if (array1[i] < array2[j]){
        result.push_back(array1[i]);
        ++i;
      } else {
        result.push_back(array2[j]);
        ++j;
      }
    }
  }
  return result;
}

